crates/project/src/project.rs

```diff
@@ -1288,27 +1292,46 @@ impl Project {
             .filter_map(move |worktree| worktree.upgrade())
     }
 
-    /// Collect all user-visible worktrees, the ones that appear in the project panel
+    /// Collect all user-visible worktrees, ordered using the worktree_order field,
+    /// the ones that appear in the project panel
     pub fn visible_worktrees<'a>(
         &'a self,
         cx: &'a AppContext,
     ) -> impl 'a + DoubleEndedIterator<Item = Model<Worktree>> {
-        self.worktrees.iter().filter_map(|worktree| {
-            worktree.upgrade().and_then(|worktree| {
-                if worktree.read(cx).is_visible() {
-                    Some(worktree)
-                } else {
-                    None
-                }
-            })
-        })
+        self.worktree_order
+            .iter()
+            .filter_map(|id| self.worktree_for_id(*id, cx))
```

Comment:
I know that there's hard to fit hundreds of worktrees inside one workspace, but I am concerned still: this looks like an `n^2` asymptotic complexity here?

Could we maybe allocate a HashMap or maybe change `worktree_order: Vec<WorktreeId>` to that `HashMap` instead?

Two methods below have a similar issue, also with id's involved.

Reply:
I'll see what I can do! It would be nice to have it as efficient as possible, but I'm wary of how effective a `HashMap` might be for ordered values 🤔 it would make the worktree lookup quicker but would be more steps to shift them around if we use it to store the order in the project itself. Though using it as an intermediary may work nicely.

Reply:
Well, it's not exactly pretty, but it works and I believe it avoids n^2 issues! 🎉